#14: Form Rendering

MHajoha opened the pull request:
Bildet die in der QPPE API spec vorhandenen form Elemente ab (inkl. JSON-(De)serialisierung) und erlaubt es, diese zu rendern.

Out of Scope: Holen der Form vom QuestionPy-Server.

Die Form-APIs von Moodle haben mir dabei einigen Kopfschmerz bereitet. Da scheint kein konsistentes Konzept dahinterzustehen:
* Die meisten Elemente fügt man mit `$mform->addElement` hinzu.
* Elemente in Gruppen erstellt man jedoch erst mit `$mform->createElement` und fügt sie in ein Array ein, welches man dann an `$mform->addElement("group", ...)` gibt.
* Der `Select all/none` checkbox controller (bei uns `CheckboxGroup`) wird gar nicht in der `\MoodleQuickForm` hinzugefügt, sondern direkt in der `\moodleform` (Superklasse von `question_edit_form`)
* Einige Optionen der Elemente gibt man im `addElement`/`createElement`-Call als Argument mit. Mal separat, mal im `$attributes`-Array
* Andere Optionen muss man im Nachhinein mithilfe von Methoden auf der Element-Instanz, die von `addElement`/`createElement` zurückgegeben wird, setzen.
* Und den erwarteten Wertetyp setzt man unabhängig vom Element mit `$mform->setType($elementName, PARAM_TEXT)`.

Um dadurch nicht 100%igen Spaghetticode schreiben zu müssen, habe ich die Unterschiede in `render_context` abstrahiert, der dann an die `\MoodleQuickForm` oder `\moodleform` deligiert. Das `group_element` verwendet für die ihm untergeordneten Element einen speziellen `group_render_context`, dessen `add_element` das Element nur in ein o.g. Array hinzufügt.

Auch das Testen des Renderings ist dadurch nicht straightforward. Ich habe mich da letzten Endes für Snapshot-Testing á la Jest entschieden.

Leider müssen für den Moodle PHPdoc Checker **alle** Funktionen und Properties ein PHPdoc mit einer Beschreibung haben. Ich habe versucht, diese viel zu verbosen Regeln zu deaktivieren, hatte aber keinen Erfolg. Die CLI von `local_moodlecheck` hat eine `--rules` Option, mit der das geht, aber `moodle-plugin-ci` nicht :(.
Deshalb habe ich den Schritt aus der CI entfernt.

Closes #14

MHajoha committed:
feat: form rendering

Closes #14

MHajoha committed:
style: follow moodle's wild code style

MHajoha committed:
style: add copyright notices

MHajoha committed:
refactor: replace collection classes with arrays and type comments

MHajoha committed:
fix: boolean element attributes

MHajoha committed:
docs: document all the things that make sense documenting

MHajoha committed:
test: explicitly include formslib in element_html_test

MHajoha committed:
style: move test_moodleform into separate file

MHajoha committed:
style: add package to PHP doc on renderable::render_to

MHajoha committed:
ci: remove PHP 7.3 from matrix

MartinGauk commented on classes/form/elements/checkbox_element.php:
```diff
@@ -0,0 +1,87 @@
+        ];
+        if ($this->selected) {
+            // FIXME: this seems to be broken within moodle, as the checked attribute never makes it into the HTML.
+            $attributes["checked"] = "checked";
```
Die defaults oder im state gespeicherten Werte sollten eventuell besser über `setDefault` gesetzt werden (statt $attributes["checked"/"value"]).

Der normale Weg in den mforms ist der, dass man die defaults mittels `setDefault` setzt. Wenn man bestehende Daten hat (aus der DB z.B.), dann werden diese mittels `set_data` außerhalb der Klasse gesetzt. Auf diese Weise will man die Definition des Formulars und die gespeicherten Daten getrennt halten.

MHajoha replied:
Eieiei, nach viel Step Debugging habe ich herausgefunden, dass der checkbox controller den Wert im Nachhinein leise überschreibt. Es ist also nicht (ohne Hacks) möglich, unterschiedliche Standardwerte innerhalb eines checkbox controllers anzugeben. :disappointed: 

Ich ändere das trotzdem wie vorgeschlagen auf `setDefault`, das klingt richtiger :+1:

MartinGauk commented on classes/form/elements/checkbox_element.php:
```diff
@@ -0,0 +1,87 @@
+            $attributes["checked"] = "checked";
+        }
+        if ($this->required) {
+            $attributes["required"] = "required";
```
Auch hier dürfte es besser sein, über `$mform->addRule` das Feld als required zu setzen. Dann wird das Feld zudem von Moodle sichtbar als required gekennzeichnet. Allerdings möchte die Funktion einen String für die Fehlermeldung erhalten.
Wäre die Frage, ob wir hier immer eine generische Fehlermeldung ausgeben lassen wollen oder ob wir unseren Fragepaketen die Möglichkeit geben wollen, eine eigene Fehlermeldung anzugeben.

MHajoha replied:
Von `question_edit_form::definition()` abgeguckt: Für `message` kann auch `null` gegeben werden, dann scheint eine Standardmessage angezeigt zu werden. In dem Fall sieht das so aus:
![image](https://user-images.githubusercontent.com/29020266/190200466-875a8551-cec3-4bac-953a-c4dcde8d71c0.png)

Eine eigene Message wäre vielleicht auch sinnvoll, würde ich jetzt aber erstmal ausklammern wollen.

MHajoha committed:
feat: use setDefault and addRule instead of attrs

MHajoha committed:
docs: make moodlecheck (almost) happy

Still needs moodlehq/moodle-local_moodlecheck#94, which may take a while to make it into moodle-plugin-ci.

MHajoha committed:
test: update element HTML snapshots

The "required" message changed on Moodle master.

MHajoha committed:
ci: run against moodle master and don't run on PR sync
